src/hotspot/os/linux/cgroupV2Subsystem_linux.hpp

```diff
@@ -39,31 +39,72 @@ class CgroupV2Controller: public CgroupController {
+                                            _cgroup_path(os::strdup(cgroup_path)),
+                                            _path(construct_path(mount_path, cgroup_path)) {
+    }
+    // Shallow copy constructor
+    CgroupV2Controller(const CgroupV2Controller& o) :
+                                            _mount_path(o._mount_path),
+                                            _cgroup_path(o._cgroup_path),
+                                            _path(o._path) {
+    }
+    ~CgroupV2Controller() {
+      // At least one controller exists with references to the paths
+    }
+
+    char *subsystem_path() override { return _path; }
+};
+
+class CgroupV2CpuController: public CgroupCpuController {
+  private:
+    CgroupV2Controller _reader;
+    CgroupV2Controller* reader() { return &_reader; }
+  public:
+    CgroupV2CpuController(CgroupV2Controller reader) : _reader(reader) {
+    }
+    int cpu_quota();
+    int cpu_period();
+    int cpu_shares();
+};
+
+class CgroupV2MemoryController final: public CgroupMemoryController {
+  private:
+    CgroupV2Controller _reader;
+    CgroupV2Controller* reader() { return &_reader; }
+  public:
+    CgroupV2MemoryController(CgroupV2Controller reader) : _reader(reader) {
     }
 
-    char *subsystem_path() { return _path; }
+    jlong read_memory_limit_in_bytes(julong upper_bound) override;
+    jlong memory_and_swap_limit_in_bytes(julong host_mem, julong host_swp) override;
+    jlong memory_and_swap_usage_in_bytes(julong host_mem, julong host_swp) override;
+    jlong memory_soft_limit_in_bytes(julong upper_bound) override;
+    jlong memory_usage_in_bytes() override;
+    jlong memory_max_usage_in_bytes() override;
+    jlong rss_usage_in_bytes() override;
+    jlong cache_usage_in_bytes() override;
+    void print_version_specific_info(outputStream* st, julong host_mem) override;
 };
 
 class CgroupV2Subsystem: public CgroupSubsystem {
   private:
     /* One unified controller */
-    CgroupController* _unified = nullptr;
+    CgroupV2Controller _unified;
     /* Caching wrappers for cpu/memory metrics */
-    CachingCgroupController* _memory = nullptr;
-    CachingCgroupController* _cpu = nullptr;
+    CachingCgroupController<CgroupMemoryController>* _memory = nullptr;
+    CachingCgroupController<CgroupCpuController>* _cpu = nullptr;
 
-    jlong mem_swp_limit_val();
-    jlong mem_swp_current_val();
+    CgroupV2Controller* unified() { return &_unified; }
 
   public:
-    CgroupV2Subsystem(CgroupController * unified) {
-      _unified = unified;
-      _memory = new CachingCgroupController(unified);
-      _cpu = new CachingCgroupController(unified);
+    CgroupV2Subsystem(CgroupV2MemoryController* memory,
+                      CgroupV2CpuController* cpu,
+                      CgroupV2Controller unified) :
+        _unified(unified),
+        _memory(new CachingCgroupController<CgroupMemoryController>(memory)),
+        _cpu(new CachingCgroupController<CgroupCpuController>(cpu)) {
```

Comment:
@jerboaa, Github doesn't let me comment on unchanged lines, so I'm not trying to talk about this line, but right beneath it :-). There are a lot of methods which are still declared which I think have been moved to the `CgroupV2MemoryController`, and so these should be removed, right?